### PR TITLE
charmcraft: 2.7.0 -> 3.1.1

### PR DESCRIPTION
**Pull request by jnsgruk**
## Description of changes

Updates Charmcraft to 3.1.1 to follow the upstream, and introduces a new Python library `craft-platforms` along the way.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by drupol — APPROVED** [review]
Diff LGTM

File: pkgs/development/python-modules/craft-platforms/default.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  src = fetchFromGitHub {
+    owner = "canonical";
+    repo = "craft-platforms";
+    rev = "refs/tags/${version}";
+    hash = "sha256-KzskmSw7NsH1CAYjPf2281Ob71Jd6AhWxtp5tR3IqyU=";
+  };
+
+  postPatch = ''
+    substituteInPlace pyproject.toml \
+      --replace-fail "setuptools>=69.0" "setuptools"
```
Comment (natsukium) [inline pkgs/development/python-modules/craft-platforms/default.nix:30]: What is this patch for?

File: pkgs/development/python-modules/craft-platforms/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    substituteInPlace craft_platforms/__init__.py \
+      --replace-fail "dev" "${version}"
+  '';
+
+  build-system = [ setuptools-scm ];
```
Comment (natsukium) [inline pkgs/development/python-modules/craft-platforms/default.nix:36]:
```suggestion
  build-system = [
    setuptools
    setuptools-scm
  ];
```
Please add `setuptools` to `build-system` explicitly.

File: pkgs/by-name/ch/charmcraft/package.nix [hunk 3]
```diff
@@ -65,19 +63,19 @@ python.pkgs.buildPythonApplication rec {
...
     requests-toolbelt
     requests-unixsocket
-    setuptools # see substituteInPlace above
     snap-helpers
     tabulate
     urllib3
   ];
 
-  build-system = with python.pkgs; [ setuptools ];
+  build-system = with python.pkgs; [ setuptools-scm ];
```
Comment (natsukium) [inline pkgs/by-name/ch/charmcraft/package.nix:71]:
```suggestion
  build-system = with python.pkgs; [
    setuptools
    setuptools-scm
  ];
```
ditto

File: pkgs/development/python-modules/craft-platforms/default.nix (hunk 1 above)
Comment (jnsgruk) [reply pkgs/development/python-modules/craft-platforms/default.nix]: Removed! Copy/pasta - sorry!

File: pkgs/development/python-modules/craft-platforms/default.nix (hunk 2 above)
Comment (jnsgruk) [reply pkgs/development/python-modules/craft-platforms/default.nix]: Done!

File: pkgs/by-name/ch/charmcraft/package.nix (hunk 3 above)
Comment (jnsgruk) [reply pkgs/by-name/ch/charmcraft/package.nix]: Done!

**Review by kirillrdy — APPROVED** [review]

**Review by drupol — APPROVED** [review]

**Pull request merged**